api/common/sharing/utils.go

```diff
@@ -12,64 +12,229 @@ import (
+		query.ByField(query.EqualsOperator, "id", instanceID),
 	)
 	if err != nil {
-		log.C(ctx).Errorf("failed retrieving service instance %s: %v", referencedInstanceID.String(), err)
-		return "", util.HandleInstanceSharingError(util.ErrReferencedInstanceNotFound, referencedInstanceID.String())
+		log.C(ctx).Errorf("failed to retrieve service instance %s: %v", instanceID, err)
+		return nil, util.HandleInstanceSharingError(util.ErrReferencedInstanceNotFound, instanceID)
 	}
-	referencedInstance := referencedInstanceObj.(*types.ServiceInstance)
+	instance := instanceObj.(*types.ServiceInstance)
 
-	referencePlan, ok := types.PlanFromContext(ctx)
-	if !ok || referencePlan == nil {
-		return "", util.HandleStorageError(util.ErrNotFoundInStorage, types.ServicePlanType.String())
+	if !instance.IsShared() {
+		log.C(ctx).Debugf("The target instance %s is not shared.", instance.ID)
+		return nil, util.HandleInstanceSharingError(util.ErrReferencedInstanceNotShared, instance.ID)
 	}
-	// verify the reference plan and the target instance are of the same service offering:
-	sameOffering, err := isSameServiceOffering(ctx, repository, referencePlan.ServiceOfferingID, referencedInstance.ServicePlanID)
+
+	sameOffering, err := isSameServiceOffering(ctx, repository, offeringID, instance.ServicePlanID)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	if !sameOffering {
-		log.C(ctx).Debugf("The target instance %s is not of the same service offering.", referencedInstance.ID)
-		return "", util.HandleInstanceSharingError(util.ErrReferenceWithWrongServiceOffering, referencedInstance.ID)
+		log.C(ctx).Debugf("The target instance %s is not of the same service offering.", instance.ID)
+		return nil, util.HandleInstanceSharingError(util.ErrReferenceWithWrongServiceOffering, instance.ID)
 	}
 
-	if !referencedInstance.IsShared() {
-		log.C(ctx).Debugf("The target instance %s is not shared.", referencedInstance.ID)
-		return "", util.HandleInstanceSharingError(util.ErrReferencedInstanceNotShared, referencedInstance.ID)
+	return instance, nil
+}
+
+func validateParameters(parameters map[string]gjson.Result) error {
+	/* failures:
+	 * parameters with empty values
+	 * combination of ID with any other selector
+	 * parameters should not be empty, but the values of each parameter can be "" (empty string/object)
+	 */
+	if len(parameters) == 0 {
+		return util.HandleInstanceSharingError(util.ErrMissingOrInvalidReferenceParameter, instance_sharing.ReferencedInstanceIDKey)
+	}
+	ID, IDExists := parameters[instance_sharing.ReferencedInstanceIDKey]
+	name, nameExists := parameters[instance_sharing.ReferenceInstanceNameSelector]
+	plan, planExists := parameters[instance_sharing.ReferencePlanNameSelector]
+	labels, labelsExists := parameters[instance_sharing.ReferenceLabelSelector]
+	var selectorLabels types.Labels
+	if err := util.BytesToObject([]byte(labels.Raw), &selectorLabels); labelsExists && err != nil {
+		return err
 	}
```

Comment:
util.BytesToObject works??
don't you need to init selectorLabels?
check if labelsExists before unmarshaling

Reply:
it works but initialized it anyway.
moved the `labelsExists` condition from the error handling to the unmarshal step.